crates/bevy_ecs/src/schedule/schedule.rs

```diff
@@ -977,6 +1025,86 @@ impl ScheduleGraph {
         Ok(self.build_schedule_inner(dependency_flattened_dag, hier_results.reachable))
     }
 
+    // modify the graph to have sync nodes for any depedants after a system with deferred system params
+    fn auto_insert_apply_deferred(
+        &mut self,
+        dependency_flattened: &mut GraphMap<NodeId, (), Directed>,
+    ) -> Result<GraphMap<NodeId, (), Directed>, ScheduleBuildError> {
+        let mut sync_point_graph = dependency_flattened.clone();
+        let topo = self.topsort_graph(dependency_flattened, ReportCycles::Dependency)?;
+
+        let get_sync_point = |graph: &mut ScheduleGraph, distance: u32| {
+            graph
+                .auto_sync_node_ids
+                .get(&distance)
+                .copied()
+                .or_else(|| {
+                    let node_id = graph.add_auto_sync();
+                    graph.auto_sync_node_ids.insert(distance, node_id);
+                    Some(node_id)
+                })
+                .unwrap()
+        };
+        dbg!(&topo);
+        // calculate the number of sync points each sync point is from the beginning of the graph
+        // use the same sync point if the distance is the same
+        let mut distances: HashMap<usize, Option<u32>> = HashMap::with_capacity(topo.len());
```

Comment:
let me know if there's a better data structure to use here. My brain didn't want to think about the problem anymore. I couldn't use the node.index() here as a index into a Vec anymore, because it's no longer dense. The auto sync ids are not included.